### PR TITLE
Issue #7730: Update AbstractChecks to log DetailAST - EmptyLineSeparator

### LINKED ISSUE #7730: Update AbstractChecks to log DetailAST - EmptyLineSeparator
Child of https://github.com/checkstyle/checkstyle/issues/5777

This issue specifically focuses on EmptyLineSeparator.

### PR DESCRIPTION
**Pull request by shashwatj07**
Closes #7730: Update AbstractChecks to log DetailAST - EmptyLineSeparator
Diff report: https://shashwat70.github.io/diff/

### CODE REVIEW

**Comment by rdiachenko** [comment]
@shashwat70 point 4 from the parent issue wasn't done. Please do and update your pull request accordingly.

**Comment by shashwat70** [comment]
> @shashwat70 point 4 from the parent issue wasn't done. Please do and update your pull request accordingly.

Done. Please review. @rdiachenko

**Comment by rdiachenko** [comment]
@shashwat70 your diff is empty but a lot of changes are expected. Looks like you didn't generate a report after your changes against EmptyLineSeparatorCheck. Please recheck config you used for diff generation.

**Comment by shashwat70** [comment]
> @shashwat70 your diff is empty but a lot of changes are expected. Looks like you didn't generate a report after your changes against EmptyLineSeparatorCheck. Please recheck config you used for diff generation.

I'm sorry for such a silly mistake, Sir. However, it's fixed now. Please review. @rdiachenko

**Comment by rdiachenko** [comment]
This check has an interesting case - https://github.com/checkstyle/checkstyle/blob/a8a90cf2423c27fb3fe227fc1906fdfc255f5d9b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java#L419. It logs messages for empty lines, but there is no AST node for an empty line.

As a possible solution, we can use the last AST node after which empty lines were found and log a single message with the number of found empty lines. @rnveach what do you think?

**Comment by rdiachenko** [comment]
@strkkk / @pbludov looks like @rnveach is quite busy, could you take a look at my concern above if you have time - https://github.com/checkstyle/checkstyle/pull/7956#issuecomment-605635983 ?

**Comment by strkkk** [comment]
@rdiachenko violation message for this case says that 
`''{0}'' has more than 1 empty lines before`
so from this point of view it totally makes sense to log it on next ast node, but I am not sure that including number of lines above is necessary.

**Comment by rnveach** [comment]
Sorry for the delay. I am trying to atleast trim my list down.

I would also get @romani's input as this is an interesting case.
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java#L419

I am not familiar with this check and would need to know more about what it is picking as violations. What lines are it picking for violations? Is it a static list or is it lines that are not empty, or .... If it is specific lines that can be described in the message, like "remove empty lines starting X lines down, and remove the next Y lines" that is an option as long as message doesn't get too complex. If it can't be described, I would want to know what lines I am expected to fix and not have to guess, so I feel we should keep logging of specific lines.

Since we are all busy during GSoC, I would put this to a new issue to review. Lets complete what we have now, which is a good start, and add to filter that it will be finalized in new issue. If we find out we can't make this check completely xpath supportive, we will just add a note to filter and check what part isn't supported.

**Comment by romani** [comment]
I recommend to move such nuance of violations that can not be xpath suppressed to Attention note of Check and create new issue ("inability to use xpath for violation on empty line" ) to decide on how to report violations in such cases.
Users can use other suppression meanwhile.

This PR should update only `log` where it is possible now without any big refactoring.

**Comment by romani** [comment]
@shashwat70 , ping

**Comment by shashwat70** [comment]
@romani I'm sorry, I got engaged with other issues and somehow missed this. I'll do the needful immediately.

**Comment by shashwat70** [comment]
@romani done!

**Comment by shashwat70** [comment]
> create new issue ("inability to use xpath for violation on empty line" ) to decide on how to report violations in such cases.
moved to #8179. Please have a look. @romani

**Comment by rnveach** [comment]
Wercker restarted as it became stuck.

**Comment by shashwat70** [comment]
> Wercker restarted as it became stuck.

Thank You. All CIs have passed. Please review.

File: src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionEmptyLineSeparatorTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+
+import com.puppycrawl.tools.checkstyle.DefaultConfiguration;
+import com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck;
+
+public class XpathRegressionEmptyLineSeparatorTest extends AbstractXpathTestSupport {
```
Comment (rdiachenko) [inline src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionEmptyLineSeparatorTest.java:32]: Could you add two more tests: one with static initializer block and another with variable definition.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 2]
```diff
@@ -44,8 +44,21 @@
  * not after line where token is found.
  * If token does not have same type sibling then empty line
  * is required at its end (for example for CLASS_DEF it is after '}').
- * Also, trailing comments are skipped.
+ * Also, trailing comments are skipped. Moreover, the violations
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:47]: Please change it to:
```
Also, trailing comments are skipped.
</p>
<p>
ATTENTION: violations from multiple empty lines cannot be suppressed via XPath (#8179).
Example:
</p>
<pre>
class A {
  void foo() {



  }
}
</pre>

File: src/xdocs/config_whitespace.xml [hunk 3]
```diff
@@ -233,8 +233,17 @@ for (Iterator foo = very.long.line.iterator();
           not after line where token is found.
           If token does not have same type sibling then empty line
           is required at its end (for example for CLASS_DEF it is after '}').
-          Also, trailing comments are skipped.
+          Also, trailing comments are skipped. Moreover, the violations
```
Comment (rdiachenko) [inline src/xdocs/config_whitespace.xml:236]: Please update according to the above comment.

**Comment by rdiachenko** [comment]
@shashwat70 ping

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/pull/8296 did it right way

**Pull request closed without merging**